governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,220 @@
+    
+    private final Provider<ScopeCleanupMarker> markerProvider = new Provider<ScopeCleanupMarker>() {
+        @Override
+        public ScopeCleanupMarker get() {
+            return new ScopeCleanupMarker();
+        }
+    };
+    
+    public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
+        this.scopeBindings = new HashMap<>(scopeBindings);
+        reqQueueExecutor.submit(new Runnable() {
+            public void run() {
+                while (true) {
+                    try {
+                        Reference<? extends ScopeCleanupMarker> ref = markerReferenceQueue.remove(500);
+                        if (ref != null && ref instanceof ScopeCleanupAction) { 
+                            Integer markerKey = ((ScopeCleanupAction)ref).getId();
+                            ScopeCleanupAction cleanupAction;
+                            synchronized(scopedCleanupActions) {
+                                cleanupAction = scopedCleanupActions.remove(markerKey);
+                            }
+                            if (cleanupAction != null) {
+                                cleanupAction.call();
+                            }
+                        }
+                    } 
+                    catch (InterruptedException e) {
+                    }
+                }
+            }
+        });
+    }
+    
+    public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
+        return binding.acceptScopingVisitor(new ManagedInstanceScopingVisitor(destroyableInstance, binding.getSource(), action));
+    }
+    
+    public void addScopeBindings(Map<Class<? extends Annotation>, Scope> bindings) {
+        scopeBindings.putAll(bindings);
+    }
+
+    /*
+     * compatibility-mode - scope is assumed to be eager singleton
+     */
+    public <T> boolean register(T destroyableInstance, Object context, Iterable<LifecycleAction> action) {
+        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
+    }
+    
+    /**
+     * final cleanup of managed instances if any
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        if (!reqQueueExecutor.isShutdown()) reqQueueExecutor.shutdown();
+        synchronized(scopedCleanupActions) {
+            for (Callable<Void> actions : scopedCleanupActions.values()) {
+                actions.call();
+            }
+            scopedCleanupActions.clear();
+        }
+        
+        for (Callable<Void> action : cleanupActions) {
+            action.call();
+        }     
+        cleanupActions.clear();
+    }
+    
+    /**
+     * visits bindingScope of managed instance to set up an appropriate strategy for cleanup
+     *  
+     */
+    private final class ManagedInstanceScopingVisitor implements BindingScopingVisitor<Boolean> {       
```

Comment:
It took me a while to figure out that ManagedInstanceScopingVisitor was adding to cleanupActions.  I believe a cleaner implementation would be to have ManagedInstanceScopingVisitor implements BindingScopingVisitor<LifecycleAction>  (instead of Boolean) and return the action object or null if there is none.  This would greatly improve testability of ManagedInstanceScopingVisitor.